### PR TITLE
[SYCLomatic][python_setup] Enabled Python build scripts migration

### PR DESCRIPTION
**Pull request by TejaX-Alaghari**
Initial PR for enabling setup.py migration framework

### CODE REVIEW

**Review by tomflinda — CHANGES_REQUESTED** [review]
@TejaX-Alaghari 
```
clang/lib/DPCT/MigratePythonSetupScript.cpp
clang/lib/DPCT/MigratePythonSetupScript.h
```
Has duplicate function's implementation  with 
```
clang/lib/DPCT/MigrateCmakeScript.cpp
clang/lib/DPCT/MigrateCmakeScript.h
```
So, I suggest that:
Create two more files, like:
```
clang/lib/DPCT/MigrateBuildScript.cpp
clang/lib/DPCT/MigrateBuildScript.cpp
```
And extract common implementation from MigratePythonSetupScript.cpp and MigrateCmakeScript.cpp to the two files above, so than the duplicate code could be reduced.

File: clang/test/dpct/cmake_migration/case_051/ref.txt [hunk 1]
```diff
@@ -1 +1 @@
-// CHECK: Warning: CMake build script file like CMakeLists.txt is not found, so no CMake build script file will be migrated.
+// CHECK: Warning: CMake or Python setup build script file like CMakeLists.txt/package.cmake or setup.py is not found, so no CMake/ Python setup build script file will be migrated.
```
Comment (tomflinda) [inline clang/test/dpct/cmake_migration/case_051/ref.txt:1]: Pls use a separate folder like py_setup_migration for python_setup migration feature, just like cmake_migration folder for cmake.

File: clang/lib/DPCT/MigratePythonSetupScript.cpp [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+static std::map<std::string /*Python setup syntax*/,
+                MetaRuleObject::PatternRewriter /*Python setup migraiton rule*/>
+    PythonSetupBuildInRules;
+
+static std::map<std::string /*file path*/,
+                std::vector<std::string> /*warning msg*/>
+    FileWarningsMap;
+
+void collectPythonSetupScripts(const clang::tooling::UnifiedPath &InRoot,
```
Comment (tomflinda) [inline clang/lib/DPCT/MigratePythonSetupScript.cpp]: The main logic is the same as https://github.com/tomflinda/SYCLomatic/blob/c1595f5689d74c1d7a4b14b37d91451b510d37f4/clang/lib/DPCT/MigrateCmakeScript.cpp#L118,  suggest to extract a common function to MigrateBuildScript.cpp/h file.

File: clang/lib/DPCT/MigratePythonSetupScript.cpp [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+      llvm::StringRef Name =
+          llvm::sys::path::filename(FilePath.getCanonicalPath());
+      if (Name.ends_with(".py")) {
+        PythonSetupScriptFilesSet.push_back(FilePath.getCanonicalPath().str());
+      }
+    }
+  }
+}
+
+bool loadBufferFromPythonSetupScriptFile(
```
Comment (tomflinda) [inline clang/lib/DPCT/MigratePythonSetupScript.cpp]: Same as above.

File: clang/lib/DPCT/MigratePythonSetupScript.cpp [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+  clang::tooling::UnifiedPath OutFileName(InFileName);
+  if (!rewriteCanonicalDir(OutFileName, InRoot, OutRoot)) {
+    return false;
+  }
+  createDirectories(path::parent_path(OutFileName.getCanonicalPath()));
+  PythonSetupScriptFileBufferMap[OutFileName] = readFile(InFileName);
+  return true;
+}
+
+bool pythonSetupScriptFileSpecified(
```
Comment (tomflinda) [inline clang/lib/DPCT/MigratePythonSetupScript.cpp]: Same as above.

File: clang/lib/DPCT/MigratePythonSetupScript.cpp [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    if (!llvm::sys::path::has_extension(FilePath) ||
+        llvm::sys::path::filename(FilePath).ends_with(".py")) {
+      IsPythonSetupScript = true;
+      break;
+    }
+  }
+  return IsPythonSetupScript;
+}
+
+void collectPythonSetupScriptsSpecified(
```
Comment (tomflinda) [inline clang/lib/DPCT/MigratePythonSetupScript.cpp]: Same as above.

File: clang/lib/DPCT/MigratePythonSetupScript.cpp [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+          PythonSetupScriptFilesSet.push_back(FilePath);
+        }
+      }
+    }
+  } else {
+    collectPythonSetupScripts(InRoot, OutRoot);
+  }
+}
+
+static void unifyInputFileFormat() {
```
Comment (tomflinda) [inline clang/lib/DPCT/MigratePythonSetupScript.cpp]: Same as above.

File: clang/lib/DPCT/MigratePythonSetupScript.cpp [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+    const std::string FileName = Entry.first.getPath().str();
+
+    // Convert input file to be LF
+    bool IsCRLF = fixLineEndings(Buffer, Buffer);
+    ScriptFileCRLFMap[Entry.first] = IsCRLF;
+  }
+}
+
+static void
+applyPythonSetupMigrationRules(const clang::tooling::UnifiedPath InRoot,
```
Comment (tomflinda) [inline clang/lib/DPCT/MigratePythonSetupScript.cpp]: Same as above.

File: clang/test/dpct/python_setup_migration/case_001/case_001_framework.cpp [hunk 8]
```diff
@@ -0,0 +1,8 @@
+// RUN: rm -rf %T && mkdir -p %T
+// RUN: cd %T
+// RUN: cp %S/python_setup_input.py ./python_setup_input.py
+// RUN: mkdir -p subdir
+// RUN: dpct -in-root ./ -out-root out  ./python_setup_input.py --migrate-build-script-only
```
Comment (tomflinda) [inline clang/test/dpct/python_setup_migration/case_001/case_001_framework.cpp]: If the in-root directory contains both *.py files and CMake script files, will SYCLomatic migrate both these two kinds of files together when "--migrate-build-script-only" is specified?
If it does, when users only want to migrate one kind of build script file, such as *py, the file type should also be specified by "--migrate-build-script=Python_Setup",  pls consider this scenario and add lit test for it.

Comment (TejaX-Alaghari) [reply clang/test/dpct/python_setup_migration/case_001/case_001_framework.cpp]: 1. Yes, "**--migrate-build-script-only**" will migrate **both CMake and Python** build script files
2. Yes, the new option to migrate only Python build script files is "--migrate-build-script=**Python_Setup**"

Comment (tomflinda) [reply clang/test/dpct/python_setup_migration/case_001/case_001_framework.cpp]: For 1, pls add a lit test for this scenario: If in-root, out-root, and migrate-build-script-only are specified, then both CMake scripts and Python scripts in the in-root directory should be migrated.
For 2, so, when "--migrate-build-script=Python_Setup" is specified, then the Python script should be migrated together with the source migration, so pls add a lit test to verify that.

File: clang/lib/DPCT/MigratePythonSetupScript.cpp (hunk 2 above)
Comment (TejaX-Alaghari) [reply clang/lib/DPCT/MigratePythonSetupScript.cpp]: Done

Introduced `collectBuildScripts` function to MigrateBuildScript.cpp

File: clang/lib/DPCT/MigratePythonSetupScript.cpp (hunk 3 above)
Comment (TejaX-Alaghari) [reply clang/lib/DPCT/MigratePythonSetupScript.cpp]: Done

Added `loadBufferFromFile` function to MigrateBuildScript.cpp

File: clang/lib/DPCT/MigratePythonSetupScript.cpp (hunk 4 above)
Comment (TejaX-Alaghari) [reply clang/lib/DPCT/MigratePythonSetupScript.cpp]: Done

Introduced `buildScriptFileSpecified` function to MigrateBuildScript.cpp

File: clang/lib/DPCT/MigratePythonSetupScript.cpp (hunk 5 above)
Comment (TejaX-Alaghari) [reply clang/lib/DPCT/MigratePythonSetupScript.cpp]: Done

Introduced `collectBuildScriptsSpecified` function to MigrateBuildScript.cpp

File: clang/lib/DPCT/MigratePythonSetupScript.cpp (hunk 6 above)
Comment (TejaX-Alaghari) [reply clang/lib/DPCT/MigratePythonSetupScript.cpp]: Done

Added `unifyInputFileFormat` function to MigrateBuildScript.cpp

File: clang/lib/DPCT/MigratePythonSetupScript.cpp (hunk 7 above)
Comment (TejaX-Alaghari) [reply clang/lib/DPCT/MigratePythonSetupScript.cpp]: Left the function as is, due to hard to extract common logic between CMake and Python Setup scripts migration

Please let me know if you have any suggestion for designing the common function

File: clang/test/dpct/python_setup_migration/case_001/case_001_framework.cpp (hunk 8 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/python_setup_migration/case_001/case_001_framework.cpp]: Added the test cases

**Review by tomflinda — APPROVED** [review]
LGTM

File: clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml [hunk 9]
```diff
@@ -0,0 +1,114 @@
+# Copyright (C) Intel Corporation
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+# See https://llvm.org/LICENSE.txt for license information.
+
+# This file will be installed to folder:
+# {dpct_install_folder}/extensions/opt_rules.
+# The rule specified in this file can be imported with commandline option:
+# --rule-file={dpct_install_folder}/extensions/opt_rules/setup_rules/setup_script_migration_rule.yaml
```
Comment (zhimingwang36) [inline clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml]:
```suggestion
# --rule-file={dpct_install_folder}/extensions/opt_rules/setup_rules/python_setup_script_migration_rule.yaml
```

File: clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+- Rule: rule_import_torch
+  Kind: PySetupRule
```
Comment (zhimingwang36) [inline clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml]: let's use full name:  PythonSetupRule

File: clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+  Priority: Fallback
+  MatchMode: Partial
+  PythonSyntax: import_torch
+  In: import torch
+  Out: |
+    import torch
+    import intel_extension_for_pytorch
+    SYCL_HOME = True
```
Comment (zhimingwang36) [inline clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml]: why SYCL_Home is a bool?

File: clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+  In: from torch.utils import cpp_extension
+  Out: from intel_extension_for_pytorch.xpu import cpp_extension
+
+- Rule: rule_torch_utils_cpp_extension
+  Kind: PySetupRule
+  Priority: Fallback
+  MatchMode: Partial
+  PythonSyntax: torch_utils_cpp_extension
+  In: torch.utils.cpp_extension
+  Out: intel_extension_for_pytorch.xpu.cpp_extension
```
Comment (zhimingwang36) [inline clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml]: what if the pytorch is stock pytroch not IPEX?  
how about prepare 2 copy of the file, one for stock pytoch,  another for IPEX?  
chnage python_setup_script_migration_rule.yaml to python_setup_script_migration_rule_pytorch.yaml  and python_setup_script_migration_rule_ipex.yaml

File: clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+- Rule: rule_utils_cpp_extension
+  Kind: PySetupRule
+  Priority: Fallback
+  MatchMode: Partial
+  PythonSyntax: utils_cpp_extension
+  In: utils.cpp_extension
+  Out: intel_extension_for_pytorch.xpu.cpp_extension
```
Comment (zhimingwang36) [inline clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml]: same as above.

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 14]
```diff
@@ -460,17 +460,21 @@ DPCT_ENUM_OPTION(
...
+    clang::dpct::opt, BuildScriptKind, BuildScript,
+    clang::dpct::DpctOptionClass::OC_Attribute,
+    DPCT_OPTION_ACTIONS(clang::dpct::DpctActionKind::DAK_Migration,
+                        clang::dpct::DpctActionKind::DAK_Analysis),
+    "migrate-build-script",
+    DPCT_OPTION_VALUES(
+        DPCT_OPTION_ENUM_VALUE("CMake", int(BuildScriptKind::BS_Cmake),
+                               "Migrate the CMake file(s).", false),
+        DPCT_OPTION_ENUM_VALUE("Python_Setup", int(BuildScriptKind::BS_PySetup),
+                               "Migrate the python setup build file(s).",
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]:
```suggestion
                               "Migrate the python setup build file(s) of PyTroch based project.",
```

File: clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml (hunk 9 above)
Comment (TejaX-Alaghari) [reply clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml]: Done

File: clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml (hunk 10 above)
Comment (TejaX-Alaghari) [reply clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml]: Done

File: clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml (hunk 11 above)
Comment (TejaX-Alaghari) [reply clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml]: Changed SYCL_HOME to inherit the CMPLR_ROOT env var from oneAPI toolkit

File: clang/include/clang/DPCT/DPCTOptions.inc (hunk 14 above)
Comment (TejaX-Alaghari) [reply clang/include/clang/DPCT/DPCTOptions.inc]: Done

File: clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml (hunk 12 above)
Comment (TejaX-Alaghari) [reply clang/tools/dpct/DpctOptRules/python_setup_script_migration_rule.yaml]: Added _ipex file in this PR
Will create another PR for _pytorch rule file

File: clang/lib/DPCT/Error.cpp [hunk 15]
```diff
@@ -191,13 +191,13 @@ void ShowStatus(int Status, std::string Message) {
   case CallIndependentToolError:
     StatusString = "Error: Call to " + Message + " failed.";
     break;
-  case MigrationErrorCMakeScriptPathInvalid:
-    StatusString = "Error: Path of CMake Script is invalid.";
+  case MigrationErrorCMakeAndPythonSetupScriptPathInvalid:
+    StatusString = "Error: Path of CMake and Python setup Script is invalid.";
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Error.cpp]:
```suggestion
    StatusString = "Error: Path of CMake or Python setup Script is invalid.";
```

File: clang/lib/DPCT/Error.cpp [hunk 16]
```diff
@@ -28,8 +28,8 @@ void ShowStatus(int Status, std::string Message) {
   case MigrationSkipped:
     StatusString = "Some migration rules were skipped.";
     break;
-  case MigrationCmakeScriptCompleted:
-    StatusString = "Migration of CMake script completed.";
+  case MigrationBuildScriptCompleted:
+    StatusString = "Migration of CMake/ Python setup script completed.";
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Error.cpp]: ### 
```suggestion
    StatusString = "Migration of CMake or Python setup script completed.";
```

File: clang/include/clang/DPCT/DPCTOptions.inc [hunk 17]
```diff
@@ -460,17 +460,22 @@ DPCT_ENUM_OPTION(
...
+    clang::dpct::DpctOptionClass::OC_Attribute,
+    DPCT_OPTION_ACTIONS(clang::dpct::DpctActionKind::DAK_Migration,
+                        clang::dpct::DpctActionKind::DAK_Analysis),
+    "migrate-build-script",
+    DPCT_OPTION_VALUES(
+        DPCT_OPTION_ENUM_VALUE("CMake", int(BuildScriptKind::BS_Cmake),
+                               "Migrate the CMake file(s).", false),
+        DPCT_OPTION_ENUM_VALUE("Python_Setup", int(BuildScriptKind::BS_PySetup),
+                               "Migrate the python setup build file(s) of "
+                               "PyTorch based project",
```
Comment (zhimingwang36) [inline clang/include/clang/DPCT/DPCTOptions.inc]:
```suggestion
                               "PyTorch based project.",
```

File: clang/lib/DPCT/Error.cpp (hunk 15 above)
Comment (zhimingwang36) [reply clang/lib/DPCT/Error.cpp]: Error: Path of build Script is invalid.

File: clang/lib/DPCT/Error.cpp [hunk 18]
```diff
@@ -191,13 +191,13 @@ void ShowStatus(int Status, std::string Message) {
...
-    StatusString = "Error: Path of CMake Script is invalid.";
+  case MigrationErrorCMakeAndPythonSetupScriptPathInvalid:
+    StatusString = "Error: Path of CMake and Python setup Script is invalid.";
     break;
-  case MigrateCmakeScriptOnlyNotSpecifed:
+  case MigrateBuildScriptOnlyNotSpecifed:
     StatusString =
         "Error: option '-migrate-build-script-only' is not specified "
-        "for CMake script migartion.";
+        "for CMake/ Python setup script migartion.";
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Error.cpp]:
```suggestion
        "for build script migration.";
```

File: clang/lib/DPCT/Error.cpp [hunk 19]
```diff
@@ -191,13 +191,13 @@ void ShowStatus(int Status, std::string Message) {
   case CallIndependentToolError:
     StatusString = "Error: Call to " + Message + " failed.";
     break;
-  case MigrationErrorCMakeScriptPathInvalid:
-    StatusString = "Error: Path of CMake Script is invalid.";
+  case MigrationErrorCMakeAndPythonSetupScriptPathInvalid:
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Error.cpp]: I think we don't need  update the enum to contains all kinds of build script,  if we add one more, t hen it needs to update again.  How about give me neutral name? like (MigrationErrorbuildScriptPathInvalid)

File: docs/dev_guide/migration/migration-rules.rst [hunk 20]
```diff
@@ -56,11 +56,14 @@ Migration rules are specified in YAML files. A single rule file may contain mult
      - Required. Specifies the priority of the rule: ``Takeover`` > ``Default`` > ``Fallback``.
        When there are rule conflicts, the rule with higher priority will take precedence.
    * - Kind
-     - ``Macro`` | ``API`` | ``Header`` | ``Type`` | ``Class`` | ``Enum`` | ``DisableAPIMigration`` | ``PatternRewriter`` | ``CMakeRule``
+     - ``Macro`` | ``API`` | ``Header`` | ``Type`` | ``Class`` | ``Enum`` | ``DisableAPIMigration`` | ``PatternRewriter`` | ``CMakeRule`` | ``PythonRule``
```
Comment (zhimingwang36) [inline docs/dev_guide/migration/migration-rules.rst:59]: PythonSetupRule
```suggestion
     - ``Macro`` | ``API`` | ``Header`` | ``Type`` | ``Class`` | ``Enum`` | ``DisableAPIMigration`` | ``PatternRewriter`` | ``CMakeRule`` | ``PythonSetupRule``
```

File: docs/dev_guide/migration/migration-rules.rst [hunk 21]
```diff
@@ -56,11 +56,14 @@ Migration rules are specified in YAML files. A single rule file may contain mult
...
      - Required. Specifies the rule type.
    * - CmakeSyntax
      - String value
      - Required. Specify the CMake syntax name that will be migrated. Use the unique name for the CMake syntax.
+   * - PythonSyntax
+     - String value
+     - Optional. Specify the Python setup syntax name that will be migrated. Use the unique name for the Python syntax.
```
Comment (zhimingwang36) [inline docs/dev_guide/migration/migration-rules.rst]: it is not python syntax? 
```suggestion
     - Optional. Specify the Python setup syntax name that will be migrated. Use the unique name for the Python setup syntax.
```

File: clang/test/dpct/cmake_migration/case_051/ref.txt (hunk 1 above)
Comment (tomflinda) [inline clang/test/dpct/cmake_migration/case_051/ref.txt]: setup.py is not necessary, take https://github.com/NVlabs/stylegan3/tree/main/torch_utils  for example, the py files are shown as follows(no setup.py is found):
```
./persistence.py
./training_stats.py
./ops/bias_act.py
./ops/conv2d_gradfix.py
./ops/conv2d_resample.py
./ops/grid_sample_gradfix.py
./ops/filtered_lrelu.py
./ops/__init__.py
./ops/upfirdn2d.py
./ops/fma.py
./misc.py
./__init__.py
./custom_ops.py

```
It is different from CMake scripts, where CMakeLists.tx is necessary, so, the warning msg for py script should be updated.

File: clang/lib/DPCT/Error.cpp (hunk 16 above)
Comment (TejaX-Alaghari) [reply clang/lib/DPCT/Error.cpp]: Changed to 

"Migration of build script(s) completed."

File: clang/test/dpct/cmake_migration/case_051/ref.txt (hunk 1 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/cmake_migration/case_051/ref.txt]: Removed "setup" word from the feature

File: docs/dev_guide/migration/migration-rules.rst (hunk 21 above)
Comment (TejaX-Alaghari) [reply docs/dev_guide/migration/migration-rules.rst]: Changed "Python setup syntax" to "Python syntax"

File: docs/dev_guide/migration/migration-rules.rst (hunk 20 above)
Comment (TejaX-Alaghari) [reply docs/dev_guide/migration/migration-rules.rst:59]: Changed "PythonSetupRule" to "PythonRule"

File: clang/test/dpct/python_migration/case_001/expected.py [hunk 22]
```diff
@@ -0,0 +1,15 @@
+from setuptools import setup, Extension
+
+import torch
+import intel_extension_for_pytorch
+import os
+SYCL_HOME = os.environ.get('CMPLR_ROOT')
```
Comment (zhimingwang36) [inline clang/test/dpct/python_migration/case_001/expected.py]: SYCL_HOME is not used?

File: clang/test/dpct/cmake_migration/case_051/ref.txt [hunk 23]
```diff
@@ -1 +1 @@
-// CHECK: Warning: CMake build script file like CMakeLists.txt is not found, so no CMake build script file will be migrated.
+// CHECK: Warning: CMake build script file like CMakeLists.txt/package.cmake or Python file is not found, so no CMake or Python build script file will be migrated.
```
Comment (zhimingwang36) [inline clang/test/dpct/cmake_migration/case_051/ref.txt]: what is package.cmake?  change to "Files with .cmake suffix"?

File: clang/test/dpct/python_migration/case_002/expected.py [hunk 24]
```diff
@@ -0,0 +1,15 @@
+from setuptools import setup, Extension
+
+import torch
+import intel_extension_for_pytorch
+import os
+SYCL_HOME = os.environ.get('CMPLR_ROOT')
```
Comment (zhimingwang36) [inline clang/test/dpct/python_migration/case_002/expected.py]: SYCL_HOME is not used?

File: clang/test/dpct/cmake_migration/case_051/ref.txt (hunk 23 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/cmake_migration/case_051/ref.txt]: Done

File: clang/test/dpct/python_migration/case_002/expected.py (hunk 24 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/python_migration/case_002/expected.py]: Changed the test case and introduced new rules

File: clang/test/dpct/python_migration/case_001/expected.py (hunk 22 above)
Comment (TejaX-Alaghari) [reply clang/test/dpct/python_migration/case_001/expected.py]: Changed the test case and introduced new rules

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**